### PR TITLE
Unified entry processing.

### PR DESCRIPTION
**Pull request by dkorolev**
Designed to be able to add a generic `ReplayDone()` method to enable persistence.

### CODE REVIEW

**Comment by dkorolev** [comment]
@mzhurovich : about 100 of new lines are comments and about 500 more are a large boilerplace unit test for SFINAE.

Thus, this effectively is a LOC-negative change, although I agree it does look bulky.

Thanks!
Dima

**Comment by dkorolev** [comment]
Persistence is coming, and it turns out soberingly simple.

After this one, the next diff is https://github.com/dkorolev/Current/compare/sherlock...dkorolev:post_sherlock.

Thanks!
Dima

File: Bricks/mq/inmemory/mq.h [hunk 1]
```diff
@@ -53,107 +53,26 @@ SOFTWARE.
...
-}  // namespace mmq
-
-template <typename MESSAGE,
-          typename CONSUMER = mmq::ProcessMessageConsumer<MESSAGE>,
-          size_t DEFAULT_BUFFER_SIZE = 1024,
-          bool DROP_ON_OVERFLOW = false>
+template <typename MESSAGE, typename CONSUMER, size_t DEFAULT_BUFFER_SIZE = 1024, bool DROP_ON_OVERFLOW = false>
 class MMQ final {
  public:
   // Type of entries to store, defaults to `std::string`.
```
Comment (mzhurovich) [inline Bricks/mq/inmemory/mq.h]: Just noticed that this comment is outdated, we don't have default now. Could you remove it pls?

File: Bricks/mq/inmemory/test.cc [hunk 2]
```diff
@@ -86,26 +91,31 @@ TEST(InMemoryMQ, DropOnOverflowTest) {
...
     mmq.PushMessage(bricks::strings::Printf("M%02d", static_cast<int>(i)));
   }
-  // Wait until at least three messages are processed by the consumer to
-  // properly get dropped messages count.
-  while (c.processed_messages_ < 2) {
-    ;  // Spin lock;
-  }
 
-  // Confirm that exactly 10 messages were dropped due to a very slow consumer.
+  // Confirm that exactly 10 messages were processed and exactly 10 were dropped due to a very slow consumer.
```
Comment (mzhurovich) [inline Bricks/mq/inmemory/test.cc]: This comment was related to a removed line: `EXPECT_EQ(c.dropped_messages_, 10u);`

File: Bricks/mq/inmemory/test.cc [hunk 3]
```diff
@@ -86,26 +91,31 @@ TEST(InMemoryMQ, DropOnOverflowTest) {
...
-  EXPECT_EQ(c.dropped_messages_, 10u);
-
-  // Wait until the rest of the queued messages are processed.
   c.delay_ms_ = 0u;
   while (c.processed_messages_ != 10u) {
-    ;  // Spin lock;
+    ;  // Spin lock.
   }
 
+  EXPECT_FALSE(c.at_least_one_message_dropped_);
```
Comment (mzhurovich) [inline Bricks/mq/inmemory/test.cc]: This looks REALLY strange. 
How I read it: I have a queue with the capacity for 10 messages, push 20 there and expect that ... not a single message is dropped?!

File: Bricks/mq/interface/interface.h [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+namespace bricks {
+namespace mq {
+
+// TODO(dkorolev): Enables three abstractions for message passing:
+//
+// 1) A generic call to `operator()`.
+//    Enabling one-, two- or three-parameter signatures, returning `bool` or `void`,
+//    and accepting entries as const- or rvalue reference parameters.
+// 2) A way to do `UpdateHead()` (external sync events or timer ticks).
```
Comment (mzhurovich) [inline Bricks/mq/interface/interface.h]: s/.//

File: Bricks/mq/interface/interface.h [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+// The signature that returns `bool` may indicate that it is done processing entries,
+// and that the framework should stop feeding it with the new ones.
+// The signature returning `void` defines a listener that will never itself make a decision to stop
+// accepting entries. It can be stopped externally (via listener scope), or detached to run forever.
+//
+// C: Accept a const reference to an entry, or require a copy of which the listener will gain ownership.
+// C.1: `void/bool operator()(const T_ENTRY& entry [, index [, total]])`.
+// C.2: `void/bool operator()(T_ENTRY&& [, index [, total]])`.
+// One usecase is to `std::move()` the received entry into a different message queue.
+// If a copy of an entry is passed in, this behavior is impossible without making a clone of the entry.
```
Comment (mzhurovich) [inline Bricks/mq/interface/interface.h]: If you have "a copy of an entry", doesn't it mean that you have "a clone"?

File: Bricks/mq/inmemory/mq.h (hunk 1 above)
Comment (dkorolev) [inline Bricks/mq/inmemory/mq.h]: Yes, removed.

File: Bricks/mq/inmemory/test.cc (hunk 2 above)
Comment (dkorolev) [inline Bricks/mq/inmemory/test.cc]: Fixed.

File: Bricks/mq/inmemory/test.cc (hunk 3 above)
Comment (dkorolev) [inline Bricks/mq/inmemory/test.cc]: Yes, I knew it looks strange, and thanks for asking the right question.

I thought of it a lot and realized it's easier for both the user and the for the interface if the fact that the message is being dropped is immediately returned to the code that is synchronously pushing the entry, not to the that is asynchronously processing them.

Reworded the comment and the test in 5845d14. PTAL.

File: Bricks/mq/interface/interface.h (hunk 4 above)
Comment (dkorolev) [inline Bricks/mq/interface/interface.h]: :+1: Good eye! Fixed and reworded.

File: Bricks/mq/interface/interface.h (hunk 5 above)
Comment (dkorolev) [inline Bricks/mq/interface/interface.h]: Reworded significantly in 5845d14. Please take another look. Thank you!

**Comment by dkorolev** [comment]
Great commennts, thank you @mzhurovich! Once you have another couple minutes ... no rush at all. :palm_tree: 

Thanks!
Dima

**Comment by mzhurovich** [comment]
@dkorolev Like the changes - much easier to understand, what's happening!
LGTM++ and merging.

**Pull request merged**